src/finality/justification/decode.rs

```diff
@@ -66,16 +79,23 @@ impl<'a> From<JustificationRef<'a>> for Justification {
 }
 
 #[derive(Copy, Clone)]
-pub struct PrecommitsRef<'a> {
-    inner: &'a [u8],
+pub enum PrecommitsRef<'a> {
+    Undecoded(&'a [u8]),
+    Decoded(&'a [Precommit]),
```

Comment:
This shouldn't be public, as otherwise someone can create a `PrecommitsRef::Undecoded` which will panic on usage.
Should use some kind of `PrecommitsRefInner`.